sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_discovery.py

```diff
@@ -0,0 +1,102 @@
+# ------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# -------------------------------------------------------------------------
+
+import os
+import time
+import dns.resolver
+from dns.resolver import NXDOMAIN, LifetimeTimeout, NoNameservers  # cspell:disable-line
+from dns.rdatatype import SRV  # cspell:disable-line
+from ._constants import DISABLE_APPCONFIGURATION_DISCOVERY, HTTPS_PREFIX
+
+
+class SRVRecord:
+    priority: int
+    weight: int
+    port: int
+    target: str
+    protocol: str = HTTPS_PREFIX
+
+    def __init__(self, answer):
+        self.priority = answer.priority
+        self.weight = answer.weight
+        self.port = answer.port
+        self.target = str(answer.target).rstrip(".")
+
+
+def find_auto_failover_endpoints(endpoint: str):
+    if os.environ.get(DISABLE_APPCONFIGURATION_DISCOVERY, "").lower() == "true":
+        return []
+    known_domain = _get_known_domains(endpoint)
+    if known_domain is None:
+        return []
+
+    origin = _find_origin(endpoint)
+    if origin is None or not _validate(known_domain, origin.target):
+        return []
+
+    srv_records = [origin] + _find_replicas(origin.target)
+    endpoints = []
+    for srv_record in srv_records:
+        if endpoint.startswith(HTTPS_PREFIX):
+            endpoint = endpoint[len(HTTPS_PREFIX) :]
+        if _validate(known_domain, srv_record.target) and srv_record.target != endpoint:
+            endpoints.append(HTTPS_PREFIX + srv_record.target)
+    return endpoints
+
+
+def _find_origin(domain):
+    uri = domain
+    if domain.startswith(HTTPS_PREFIX):
+        uri = domain[len(HTTPS_PREFIX) :]
+    request = f"_origin._tcp.{uri}"
+    srv_records = _request_record(request)
+    if not srv_records:
+        return None
+    return SRVRecord(srv_records[0])
+
+
+def _find_replicas(origin):
+    replicas = []
+    i = 0
+    while True:
+        request = f"_alt{str(i)}._tcp.{origin}"
+        answers = _request_record(request)
+        if not answers:
+            break
+        for answer in answers:
+            replicas.append(SRVRecord(answer))
+        i += 1
+    return replicas
+
+
+def _request_record(request):
+    now = time.time()
+    while time.time() - now < 5:
```

Comment:
I can't find a record of a conversation, but I thought it came up as an acceptable period to attempt retry.

Part of this is also that this is a sync request, so this will be blocking the rest of the application.